sn_networking/src/cmd.rs

```diff
@@ -275,9 +279,39 @@ pub struct SwarmLocalState {
 }
 
 impl SwarmDriver {
+    #[allow(clippy::mutable_key_type)] // for Bytes in NetworkAddress
+    fn select_new_replications(
+        &mut self,
+        holder: PeerId,
+        incoming_keys: &mut HashMap<NetworkAddress, RecordType>,
+    ) {
+        if self.seen_replications.len() / 2 > K_VALUE.into() {
+            self.seen_replications.retain(|_k, (time_stamp, _keys)| {
+                time_stamp.elapsed() < SEEN_REPLICATION_FADED_OUT_S
+            });
+        }
+
+        let (time_stamp, existing_keys) = self
+            .seen_replications
+            .entry(holder)
+            .or_insert((Instant::now(), Default::default()));
+        incoming_keys.retain(|k, v| match existing_keys.get(k) {
+            Some(existing_v) => v != existing_v,
+            None => true,
+        });
```

Comment:
I think we iterate through the entire `incoming_key` even if `holder` does not exist. Maybe we can match on the `Entry::Occupied` to perform `retain` operation and ignore when `Entry::Vacant`? Something like this,


```rust
let entry = self.seen_replications.entry(holder);
if let Entry::Occupied(mut occupied_entry) = entry {
    let (time_stamp, existing_keys) = occupied_entry.get_mut();
    incoming_keys.retain(|k, v| match existing_keys.get(k) {
        Some(existing_v) => v != existing_v,
        None => true,
    });
    existing_keys.extend(incoming_keys.clone());
    *time_stamp = Instant::now();
} else {
    entry.or_insert((Instant::now(), incoming_keys));
}
```